MobileSdk/src/main/java/com/spruceid/mobile/sdk/GattClient.kt

```diff
@@ -333,101 +348,300 @@ class GattClient(private var callback: GattClientCallback,
+
+                            try {
+                                btAdapter?.cancelDiscovery()
+                            } catch (e: SecurityException) {
+                                reportLog("Unable to cancel discovery.")
+                            }
+
+                            val connectThread: Thread = object : Thread() {
+                                override fun run() {
+                                    try {
+                                        // createL2capChannel() requires/initiates pairing, so we have to use
+                                        // the "insecure" version.  This requires at least API 29, which we did
+                                        // check elsewhere (we'd never have got this far on a lower API), but
+                                        // the linter isn't smart enough to know that, and we have PR merging
+                                        // gated on a clean bill of health from the linter...
+                                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                                            l2capSocket = device.createInsecureL2capChannel(channelPSM)
+                                            l2capSocket?.connect()
+                                        }
+                                    } catch (e: IOException) {
+                                        reportError("Error connecting to L2CAP socket: ${e.message}")
+                                        return
+                                    } catch (e: SecurityException) {
+                                        reportError("Not authorized to connect to L2CAP socket.")
+                                        return
+                                    }
+
+                                    l2capWriteThread = Thread { writeResponse() }
+                                    l2capWriteThread!!.start()
+
+                                    // Let the app know we're connected.
+                                    //reportPeerConnected()
+
+                                    // Reuse this thread for reading
+                                    readRequest()
+                                }
+                            }
+                            connectThread.start()
+                        }
+                    }
+                }
+
+                else -> {
+                    dprint("Unknown Changed: ${value.size}")
                 }
             }
         }
     }
 
     /**
-     *
+     * Thread for reading the request via L2CAP.
      */
-    private fun afterIdentObtained(gatt: BluetoothGatt) {
-        try {
-            // Use L2CAP if supported by GattServer and by this OS version
-            usingL2CAP = characteristicL2CAP != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q
-
-            if (usingL2CAP) {
-                callback.onLog("Using L2CAP: $usingL2CAP")
+    private fun readRequest() {
```

Comment:
Can't you have an executor instance in the class and get rid of this while loop?

Reply:
Not cleanly, I think?  This is being executed as a thread, and we've got two things going on; we have a receive loop assembling data as it arrives, and we've got a check we perform a half second after every receive to see if anything else came in after it.
The while loop is there to accumulate the data.  The executor is there to try and figure out if we have a complete message or not.
I'll add some comments to explain in the code, but the problem is that we're receiving a blob of data of unknown size; there's nothing in the protocol that lets us determine how much data will be coming in.  The data comes in as (potentially very small) chunks, and the only way we "know" we have the whole message is when no more of it arrives.
So, the executor bit is firing off a check with a half second delay (along with some timestamping) to see if we have any subsequent input.  When we get a full half-second with no data, we assume things are good and the message has arrived.